### ISSUE TITLE
Null pointer access in spdk_vhost_vq_avail_ring_get

### ISSUE THREAD

**Issue by mhae** [spdk/spdk#687]
I was starting my test VM while a vhost user scsi tgt was being attached to vhost. This lead to the following seg fault:

```
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `app/vhost/vhost -S /var/tmp -s 1024 -m 0x03'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00005586d2e9f02b in spdk_vhost_vq_avail_ring_get (virtqueue=0x20002cd10bc0, reqs=0x7ffec6a8fde2, reqs_len=1) at vhost.c:185
185             uint16_t last_idx = vring->last_avail_idx, avail_idx = avail->idx;
[Current thread is 1 (Thread 0x7f72d154a980 (LWP 5746))]
(gdb) p avail
$1 = (struct vring_avail *) 0x0
(gdb) bt
#0  0x00005586d2e9f02b in spdk_vhost_vq_avail_ring_get (virtqueue=0x20002cd10bc0, reqs=0x7ffec6a8fde2, reqs_len=1) at vhost.c:185
#1  0x00005586d2ea359c in eventq_enqueue (svsession=0x20002cd10ac0, scsi_dev_num=7, event=1, reason=1) at vhost_scsi.c:252
#2  0x00005586d2ea55df in spdk_vhost_scsi_session_add_tgt (vdev=0x20002cd1be80, vsession=0x20002cd10ac0, ctx=0x7) at vhost_scsi.c:940
#3  0x00005586d2ea0c14 in spdk_vhost_event_async_foreach_fn (arg1=0x5586d385fde0, arg2=0x7) at vhost.c:919
#4  0x00005586d2ebce85 in _spdk_event_queue_run_batch (reactor=0x5586d381efc0, thread=0x5586d381f230) at reactor.c:162
#5  0x00005586d2ebd11e in _spdk_reactor_run (arg=0x5586d381efc0) at reactor.c:245
#6  0x00005586d2ebd42f in spdk_reactors_start () at reactor.c:331
#7  0x00005586d2ebbcbb in spdk_app_start (opts=0x7ffec6a900a0, start_fn=0x5586d2e400b9 <vhost_started>, arg1=0x0, arg2=0x0) at app.c:677
#8  0x00005586d2e401aa in main (argc=7, argv=0x7ffec6a90228) at vhost.c:106
```

For whatever reason, vring_avail was NULL leading to the core.

Note, I'm using qemu **2.12** on debian that I've built from the source (basically a backport from buster to stretch). 

## Expected Behavior
Attaches while a VM comes up should not cause an abort in vhost app.

## Current Behavior
Core dump.

## Steps to Reproduce
1. SPDK from master
2. app/vhost/vhost -S /var/tmp -s 1024 -m 0x03
3. ./scripts/rpc.py construct_vhost_scsi_controller --cpumask 0x1 vhost.0
4. add/remove stress test
```
while [ 1 -eq 1 ] ; do
for i in {0..7} ; do
  ./scripts/rpc.py construct_malloc_bdev 1 512 -b Malloc$i
  ./scripts/rpc.py add_vhost_scsi_lun vhost.0 $i Malloc$i
  sleep 0.1
done
for i in {0..7} ; do
  ./scripts/rpc.py remove_vhost_scsi_target vhost.0 $i
  ./scripts/rpc.py delete_malloc_bdev Malloc$i
done
done
```
5. Start VM
```
kvm -enable-kvm -machine pc-i440fx-2.1,accel=kvm,usb=off -smp 1 -cpu SandyBridge -m 512M -object memory-backend-file,id=mem,size=512M,mem-path=/dev/hugepages/vm,share=on -numa node,memdev=mem -drive file=/home/mhae/min.qcow2,if=none,id=disk -device ide-hd,drive=disk,bootindex=0 -chardev socket,id=char0,path=/var/tmp/vhost.0 -device vhost-user-scsi-pci,id=scsi0,chardev=char0 --curses
```

Note, it takes some time to hit that condition. I believe I've also seen this null de-reference in other scenarios when I did a reboot from within the guest (reboot command).

## Context (Environment including OS version, SPDK version, etc.)
root@mhae-zbook:/home/mhae/src/spdk# uname -a
Linux mhae-zbook 4.9.0-8-amd64 #1 SMP Debian 4.9.144-3.1 (2019-02-19) x86_64 GNU/Linux
root@mhae-zbook:/home/mhae/src/spdk# kvm -version
QEMU emulator version 2.12.0 (Debian 1:2.12+dfsg-3~hpebpo9+1)
Copyright (c) 2003-2017 Fabrice Bellard and the QEMU Project developers

**Comment by pawelwod** [comment]
I try but can't reproduce this. It was running for several minutes with no issue.
Can I ask you to try building spdk with `./configure --enable-debug` and run it under gdb? Then pls attach full console logs.

I see that you are using QEMU from Debian repo. As I don't have any Debian host right now can you try https://github.com/spdk/qemu/tree/spdk-2.12 ?

**Comment by mhae** [comment]
@pawelwod The stack traces are different. Not sure it's a dup.

**Comment by mhae** [comment]
@pawelwod I'll give it a try. Might not happen today.

**Comment by mhae** [comment]
@pawelwod Please note that you have to repeatedly start the VM while the target is being attached. It can take multiple starts of the VM before this condition is hit. I think I also hit it another time when I rebooted the VM (just logged into the guest and executed the reboot command). 
I tried to reproduce this today again but wasn't successful but I've hit it a couple of times so far. I still have one core dump (w/o debug) if that helps but from what I've seen in the core, the avail ptr is NULL and we try to de-reference it. I'm wondering if a proactive check for NULL would help here.

**Comment by mhae** [comment]
Ok, I just reproduce the other issue where it aborts in spdk_vhost_scsi_session_add_tgt. I basically was running my script that adds and removes devices. While this script is running, I repeatedly started my VM and killed it (kill command) when I saw that the system was up. It took me ~10 tries to get to the core.

Thread 4 (Thread 0x7fdc6b2bf700 (LWP 26299)):
#0  0x00007fdc74ac78bd in poll () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x0000560edf7dd4ee in fdset_event_dispatch (arg=0x560edfac8fa0 <vhost_user+8192>) at fd_man.c:248
#2  0x00007fdc74d8e4a4 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#3  0x00007fdc74ad0d0f in clone () from /lib/x86_64-linux-gnu/libc.so.6

Thread 3 (Thread 0x7fdc749e7700 (LWP 26294)):
#0  0x00007fdc74ad1303 in epoll_wait () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x0000560edf83a83a in eal_intr_handle_interrupts (pfd=5, totalfds=1) at /home/mhae/src/spdk/dpdk/lib/librte_eal/linuxapp/eal/eal_interrupts.c:853
#2  0x0000560edf83aa1b in eal_intr_thread_main (arg=0x0) at /home/mhae/src/spdk/dpdk/lib/librte_eal/linuxapp/eal/eal_interrupts.c:937
#3  0x0000560edf8498f9 in rte_thread_init (arg=0x560ee0c6aa00) at /home/mhae/src/spdk/dpdk/lib/librte_eal/common/eal_common_thread.c:163
#4  0x00007fdc74d8e4a4 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#5  0x00007fdc74ad0d0f in clone () from /lib/x86_64-linux-gnu/libc.so.6

Thread 2 (Thread 0x7fdc741e6700 (LWP 26295)):
#0  0x0000560edf80df10 in spdk_thread_poll (thread=0x7fdc640008c0, max_msgs=0, now=7502552496664) at thread.c:445
#1  0x0000560edf7ed1b3 in _spdk_reactor_run (arg=0x560ee0c6f080) at reactor.c:247
#2  0x0000560edf831287 in eal_thread_loop (arg=0x0) at /home/mhae/src/spdk/dpdk/lib/librte_eal/linuxapp/eal/eal_thread.c:153
#3  0x00007fdc74d8e4a4 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#4  0x00007fdc74ad0d0f in clone () from /lib/x86_64-linux-gnu/libc.so.6

Thread 1 (Thread 0x7fdc76058980 (LWP 26293)):
#0  0x0000560edf7d54aa in spdk_vhost_scsi_session_add_tgt (vdev=0x20002cd1bc40, vsession=0x20002cd10400, ctx=0x2) at vhost_scsi.c:912
#1  0x0000560edf7d1cf7 in spdk_vhost_external_event_foreach_continue (vdev=0x20002cd1bc40, vsession=0x20002cd10400, fn=0x560edf7d5444 <spdk_vhost_scsi_session_add_tgt>, arg=0x2) at vhost.c:1326
#2  0x0000560edf7d1e17 in spdk_vhost_dev_foreach_session (vdev=0x20002cd1bc40, fn=0x560edf7d5444 <spdk_vhost_scsi_session_add_tgt>, arg=0x2) at vhost.c:1356
#3  0x0000560edf7d59c4 in spdk_vhost_scsi_dev_add_tgt (vdev=0x20002cd1bc40, scsi_tgt_num=2, bdev_name=0x560ee0c71fd0 "Malloc2") at vhost_scsi.c:1018
#4  0x0000560edf7d254b in spdk_rpc_add_vhost_scsi_lun (request=0x560ee0ca9220, params=0x560ee0c71da0) at vhost_rpc.c:148
#5  0x0000560edf7f6ddd in spdk_jsonrpc_handler (request=0x560ee0ca9220, method=0x560ee0c71d80, params=0x560ee0c71da0) at rpc.c:87
#6  0x0000560edf7f0718 in spdk_jsonrpc_server_handle_request (request=0x560ee0ca9220, method=0x560ee0c71d80, params=0x560ee0c71da0) at jsonrpc_server_tcp.c:223
#7  0x0000560edf7ef348 in parse_single_request (request=0x560ee0ca9220, values=0x560ee0c71d60) at jsonrpc_server.c:98
#8  0x0000560edf7ef76e in spdk_jsonrpc_parse_request (conn=0x7fdc6b300300, json=0x7fdc6b300318, size=137) at jsonrpc_server.c:185
#9  0x0000560edf7f093b in spdk_jsonrpc_server_conn_recv (conn=0x7fdc6b300300) at jsonrpc_server_tcp.c:285
#10 0x0000560edf7f0e30 in spdk_jsonrpc_server_poll (server=0x7fdc6b2c0010) at jsonrpc_server_tcp.c:436
#11 0x0000560edf7f7240 in spdk_rpc_accept () at rpc.c:205
#12 0x0000560edf7ed780 in spdk_rpc_subsystem_poll (arg=0x0) at rpc.c:51
#13 0x0000560edf80dfd1 in spdk_thread_poll (thread=0x560ee0c6f230, max_msgs=0, now=7502552367089) at thread.c:492
#14 0x0000560edf7ed1b3 in _spdk_reactor_run (arg=0x560ee0c6efc0) at reactor.c:247
#15 0x0000560edf7ed4af in spdk_reactors_start () at reactor.c:331
#16 0x0000560edf7ebd3b in spdk_app_start (opts=0x7ffcb78eb0d0, start_fn=0x560edf7700b9 <vhost_started>, arg1=0x0, arg2=0x0) at app.c:677
#17 0x0000560edf7701aa in main (argc=7, argv=0x7ffcb78eb258) at vhost.c:106

**Comment by pawelwod** [comment]
> I repeatedly started my VM and killed it (kill command) when I saw that the system was up

Now I have the repro :) Fixing this issue.

**Comment by mhae** [comment]
@pawelwod Sweet! Feel free to send me a patch for testing if you want.

**Comment by mhae** [comment]
Just a quick update: I'm hitting this issue quite frequently in our environment (outside of the usual SPDK build) and SeaBIOS 1.12 (latest version). I had to switch to 1.12 since the combination of SeaBIOS 1.10 and qemu 2.12.0 is causing another issue where after adding/removing a large number of targets, the start-up times after a reboot are quite long (stuck in BIOS for many minutes).

I'm not sure why I'm seeing this issue (null pointer in avail) mostly in our environment. Maybe because we're specifying more devices (incl. networking) compared to the repro above with KVM.

Anyway, a simple null pointer check for avial actually solves this issue.  Just wanted to mention this.

**Comment by pawelwod** [comment]
I'm still working on this. As a workaround for you can pull these patches

https://review.gerrithub.io/#/c/spdk/spdk/+/446627/2/ (it must be version 2 of this patch)
https://review.gerrithub.io/#/c/spdk/spdk/+/446626/1 (it must be version 1 of this patch)

They are not ready to be merged but at least the make this issue go away on my test setup.

>  I had to switch to 1.12 since the combination of SeaBIOS 1.10 and qemu 2.12.0 is causing another issue where after adding/removing a large number of targets, the start-up times after a reboot are quite long (stuck in BIOS for many minutes).

Is version 1.10 in official version in QEMU repo? If yes, it is worth opening another issue.

**Comment by mhae** [comment]
@pawelwod Sorry, didn't want to rush you ... I was just adding some more info. I looked at the patches and it seems quite some work.

Re: SeaBIOS. It seems that it comes bundled with QEMU but we use the SeaBIOS debian package. The qemu project updated the SeaBIOS to 1.11.1 in qemu 2.12 (master is at 1.12).
But remember, the official Linux distros are pretty far behind in regards to qemu.

**Comment by pawelwod** [comment]
Ok, I have the final fix (i think so :P) Here is a top of the patch series you give a try:

https://review.gerrithub.io/#/c/spdk/spdk/+/446627/

You need to pull [all of them](https://review.gerrithub.io/#/q/topic:issue_687+(status:open+OR+status:merged))

BTW: I suspect that the issue with long boot on QEMU 2.12 & SeaBIOS 1.10 might also be fixed here but just a blind shot.

**Comment by mhae** [comment]
@pawelwod Awesome! The patches fixed the issue I've been running into in my repro environment. Unfortunately, I can't try it our target environment since we're still on 18.10. I need to spend some time pulling in SPDK top of tree to see that it fixes all the issues we've seen in our target env. But the initial testing looks very promising. I did VM starts & reboots for about an hour without any glitch. Thanks for all your effort!

**Comment by pawelwod** [comment]
If I manage pass the review I think we will want to cherry pick this fix to incoming 19.01.1

**Comment by pawelwod** [comment]
There is a alternate fix pushed for review from one of the mainteners here https://review.gerrithub.io/#/c/spdk/spdk/+/448229/ but looks like is not mature enough and still have problems. Lets we wait a while till situation is cleared.

**Comment by mhae** [comment]
Sounds good. I’m happy to redo my testing. Let me know when the patch is mature enough.

**Comment by pawelkax** [comment]
Test for this issue is prepared in https://review.gerrithub.io/#/c/spdk/spdk/+/447060/ .

**Comment by darsto** [comment]
The patch from @pawelwod tried to fix too many issues at once and I was not able to review it.

In order to try to understand that patch, I started looking into the vhost code and tried to find the specific scenario that's broken. I managed to find at least 8 (sic!) separate data races, for which I submitted around ~10-15 patches. Right now we still have 4 vhost patches [1] and 2 bdev patches [2] outstanding. The vhost ones are fairly straightforward and should be merged any day now, but the bdev ones still need to be discussed - there might be better ways to fix the bdev layer.

I did not try to fix this specific github issue (or #680), but rather tried to fix all the possible data races in vhost. With all those fixes applied, I ran various stress tests like the one you've provided and was not able to crash vhost or make any RPC hang.

If possible, could you try checking out at the vhost series, cherry-picking the bdev patches and then running your tests?

D.

[1] (top of the outstanding patch series) https://review.gerrithub.io/c/spdk/spdk/+/449990
[2] https://review.gerrithub.io/c/spdk/spdk/+/450112

**Comment by mhae** [comment]
Hey Darek, just to clarify, is it enough to apply just the two patches [1]
and [2] or is there more?

**Comment by mhae** [comment]
I've been running my stress test all morning and didn't hit the issue anymore! I'm using the latest from today (latest hash is bb488d2829a9b7863daab45917dd2174905cc0ae). It seems that patch 449990 did the trick.

**Comment by darsto** [comment]
@mhae I'm finally back at the office :)

I'm glad it worked for you!
If you're running a debug build of vhost, you might still randomly hit an assert() on scsi lun hotremove, but the non-debug builds of current spdk master already work fine.

The bdev patches I linked before are still pending on gerrithub, but they only remove that failing assert now.

**Comment by darsto** [comment]
Since you've reported it fixes the original problem, I'm actually closing issue. Thanks!

The assert() on hotremove has its own github issue #733

**Comment by darsto** [comment]
Just for future reference, here's the list of all the related vhost fixes.

[vhost: ignore sessions that weren't ever started](https://review.gerrithub.io/c/spdk/spdk/+/449394/6)(Merged)
[vhost/scsi: fix newly attached targets being hotremoved by mistake](https://review.gerrithub.io/c/spdk/spdk/+/449990/4)(Merged)
[vhost/scsi: don't send hotremove sense codes after session restart](https://review.gerrithub.io/c/spdk/spdk/+/449709/6)(Merged)
[vhost/scsi: fix starting a session while SCSI targets are hotplugged](https://review.gerrithub.io/c/spdk/spdk/+/450052/3)(Merged)
[vhost/scsi: ignore hotremoved devices in newly started sessions](https://review.gerrithub.io/c/spdk/spdk/+/449393/6)(Merged)
[vhost/scsi: don't hotplug targets to stopped sessions](https://review.gerrithub.io/c/spdk/spdk/+/449392/2)(Merged)
[vhost/scsi: forbid removing targets that are still being added](https://review.gerrithub.io/c/spdk/spdk/+/449391/2)(Merged)
[vhost-scsi: separatate scsi device state from scsi session state](https://review.gerrithub.io/c/spdk/spdk/+/447072/8)(Merged)
[vhost/scsi: remove hotremoved scsi targets on device stop](https://review.gerrithub.io/c/spdk/spdk/+/439445/4)(Merged)
[vhost: change vsession-&gt;lcore only within that lcore](https://review.gerrithub.io/c/spdk/spdk/+/448229/5)(Merged)
[vhost: add spdk_vhost_session_start_done/stop_done](https://review.gerrithub.io/c/spdk/spdk/+/448228/5)(Merged)
[vhost: call session_event_done() always under the global vhost lock](https://review.gerrithub.io/c/spdk/spdk/+/448227/5)(Merged)

**Comment by darsto** [comment]
I have just found one more case where this happens. Here's the fix:
[vhost/scsi: don't send events when eventq is unset](https://review.gerrithub.io/c/spdk/spdk/+/451789)
